mandible/metadata_mapper/key.py

```diff
@@ -0,0 +1,31 @@
+from dataclasses import dataclass
+from typing import Any, List
+
+
+@dataclass(frozen=True)
+class Key:
+    key: str
+
+    # Optional arguments
+    return_list: bool = False
+    return_first: bool = False
+
+    def __post_init__(self):
+        if self.return_list and self.return_first:
+            raise ValueError(
+                "cannot set both 'return_list' and 'return_first' to True",
+            )
+
+    def resolve_list_match(self, values: List[Any]) -> Any:
+        if self.return_list:
+            return values
+
+        if not values:
+            raise KeyError(self.key)
+
+        if not self.return_first and len(values) != 1:
+            raise ValueError(
+                "returned multiple values (try setting return_list=True)",
+            )
+
+        return values[0]
```

Comment:
To me it seems better to simplify this interface. If the data is a list and you did not explicitly ask for it to be changed, give you whats in the file? 

```suggestion
from dataclasses import dataclass
from typing import Any, List


@dataclass(frozen=True)
class Key:
    key: str

    # Optional arguments
    return_first: bool = False

    def resolve_list_match(self, values: List[Any]) -> Any:
        # I feel like strictly greater than 1 makes more sense here
        if not self.return_first and len(values) < 1
            return values

        if not values:
            raise KeyError(self.key)

        return values[0]
```

Reply:
You are misunderstanding how this function is used to return data from the query. To be clear, it is only called when using jsonpath or xpath. That's because jsonpath and xpath **always return a list of nodes matching the query expression**.

For example, if this is my data:
```json
{
  "items": [
    {
      "name": "apple",
      "type": "fruit"
    },
    {
      "name": "banana",
      "type": "fruit"
    }
  ]
}
```

And this is my jsonpath expression: `item[0].name`

Then the result of the variable that is fed into this function which looks like this:
```python
values = [match.value for match in expr.find(data)]
```
Looks like this:
```python
print(values)
["apple"]
```

We know that this particular expression will only ever match 1 node of the input, but that's not the case in general.

Imagine my jsonpath expression is `item[*].name`

Then my `values` variable looks like this:
```python
print(values)
["apple", "banana"]
```

In this case we expect our query to return multiple items. These sorts of queries were simply not possible before in mandible as we didn't write any code to handle this case. We would just return the first node that matched the query. 

Note that the data that we're querying for isn't actually a list. We are trying to get the item name which is a string. However, depending on how we write the query, there might be multiple items in the data that match our expression. That's what the `return_list` variable is letting you control. It is a way for you to tell mandible whether you expect your query expression to match multiple nodes or just one. 

There also is no general way to tell from the expression whether it will match more than one node, as it depends on the data. Consider a filter query like `additional_attributes[?name = 'LookDirection'].value`. If we know that the data we're querying is supposed to have unique names for each additional attribute, then this query should only return one value. However, it could return multiple if the data is non-conformant, in which case we would want mandible to throw an error.

`return_first` is really just something I threw in to make it possible to replicate the old behavior, although I don't really expect it to be useful.

Reply:
I appreciate the detailed response.